openedx/openedx-platform
# Index Studio content using Meilisearch [FC-0040]

## Description

This is a **discovery prototype** to learn about Meilisearch and how it could be used with Open edX.

The goal is to have this available for Studio search in Redwood, but disabled by default. Interested users can enable it and give us feedback, particularly on the use of Meilisearch vs. ElasticSearch.

## Supporting information

The goal would be to implement [this planned **Studio** courseware search UX](https://www.figma.com/file/DmUC2e7BAoQsGDq4KgApPa/Content-Tagging-MVP?type=design&node-id=13343-23509&mode=design&t=VbZ41Idprq3JVHYe-0) using Meilisearch:

![Screenshot 2024-02-28 at 2 08 52 PM](https://github.com/openedx/edx-platform/assets/945577/a46b2899-286f-4a97-bc0c-9b7889d245f6)

For now, this is a ~~proof of concept~~ initial backend implementation that indexes draft (Studio) content from courses and libraries (v2).

## Screenshot

Here is [a rudimentary search UI in the course authoring MFE searching courseware using this Meilisearch index](https://github.com/openedx/frontend-app-course-authoring/pull/868):

![Screenshot 2024-03-05 at 12 15 29 PM](https://github.com/openedx/edx-platform/assets/945577/4711a15f-3e81-4946-85aa-e1051c10298d)

This shows the resulting index in the UI built in to Meilisearch:

![meilisearch-example](https://github.com/openedx/edx-platform/assets/945577/c712feb8-101b-4de3-819b-36f9a4adfc3c)

And here is an example search result with typo tolerance:

![Screenshot 2024-02-28 at 2 12 54 PM](https://github.com/openedx/edx-platform/assets/945577/03ec0053-e69c-4b7c-a338-d0346d2c5dcc)

## Testing instructions

1. Install [tutor-contrib-meilisearch](https://github.com/open-craft/tutor-contrib-meilisearch) onto a Tutor Nightly Devstack.
2. Run `tutor dev run cms bash` and `./manage.py cms reindex_libraries`
3. View the resulting index at http://meilisearch.local.edly.io:7700/ (see `tutor-contrib-meilisearch` README for how to get the API key to log in)
4. Or set up [this PR](https://github.com/openedx/frontend-app-course-authoring/pull/868) in the course authoring MFE and follow it's instructions to see the basic search UI in the MFE.

## Faceted Search Results (backend)

You can see faceted search using tags using the [frontend demo](https://github.com/openedx/frontend-app-course-authoring/pull/868). But you can also see the data on the backend if you want: From the CMS django shell (`./manage.py cms shell`), run this code to see an example faceted search:

```
from django.conf import settings
import meilisearch
client = meilisearch.Client(settings.MEILISEARCH_URL, settings.MEILISEARCH_API_KEY)
index_name = settings.MEILISEARCH_INDEX_PREFIX + "studio_content"

client.index(index_name).search("hyperspace", {"facets": ["block_type", "tags"]})
```

result:

```
 'facetDistribution': {'block_type': {'html': 2,
   'problem': 2,
   'vertical': 1,
   'video': 1},
  'tags': {},
  'tags.taxonomy': {'ESDC Skills and Competencies': 1, 'Lightcast Open Skills Taxonomy': 3}
  'tags.level0': {'ESDC Skills and Competencies > Knowledge': 1,
   'Lightcast Open Skills Taxonomy > Administration': 2,
   'Lightcast Open Skills Taxonomy > Engineering': 1},
  'tags.level1': {'ESDC Skills and Competencies > Knowledge > Physical Sciences Sub-Category': 1,
   'Lightcast Open Skills Taxonomy > Administration > Administrative Support': 1,
   'Lightcast Open Skills Taxonomy > Administration > Dictation': 1,
   'Lightcast Open Skills Taxonomy > Administration > Document Management': 1,
   'Lightcast Open Skills Taxonomy > Engineering > Aerospace Engineering Subcategory': 1},
  'tags.level2': {'ESDC Skills and Competencies > Knowledge > Physical Sciences Sub-Category > Physical Sciences': 1,
   'Lightcast Open Skills Taxonomy > Administration > Administrative Support > Administrative Functions': 1,
   'Lightcast Open Skills Taxonomy > Administration > Dictation > Transcribing': 1,
   'Lightcast Open Skills Taxonomy > Engineering > Aerospace Engineering Subcategory > Aerospace Engineering': 1,
   'Lightcast Open Skills Taxonomy > Engineering > Aerospace Engineering Subcategory > Space Exploration': 1,
   'Lightcast Open Skills Taxonomy > Engineering > Aerospace Engineering Subcategory > Space Flight': 1,
   'Lightcast Open Skills Taxonomy > Engineering > Aerospace Engineering Subcategory > Spacecraft Propulsion': 1},
  'tags.level3': {},
},
```

## Not Implemented / TODO

- [x] Add the course/library **name** to the index
- [x] Add the hierarchy to each block - the breadcrumbs of what course/section/subsection/unit it's in
- [x] Add ADR
- [ ] Restrict the search results based on what permissions the user has. Most of the code to do this is already in there; already the frontend is using a user-specific API key that is setup with restrictions on what it can search. We just need to populate those restrictions with a list of orgs/courses/libraries they can see.
    * Will do in a future PR. For now, this is disabled by default and if enabled, is only available to global staff users.

## Deadline

We're hoping to include a beta version of this (off by default) in Redwood.

Private ref: FAL-3689
